### PR TITLE
Handle newly added property sub-pattern operation in OperationExtensi…

### LINKED ISSUE #40499: Unhandled declaration pattern assert for CSharpLazyPropertySubpatternOperation in OperationExtensions
`Microsoft.CodeAnalysis.Operations.CSharpLazyPropertySubpatternOperation` generated for 'left' in below code:

```csharp
op is { LeftOperand: IBinaryOperation left, RightOperand: IBinaryOperation right }
```

```
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.OperationExtensions.GetValueUsageInfo(Microsoft.CodeAnalysis.IOperation operation) Line 95	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.Walker.OnReferenceFound(Microsoft.CodeAnalysis.ISymbol symbol, Microsoft.CodeAnalysis.IOperation operation) Line 101	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.Walker.VisitDeclarationPattern(Microsoft.CodeAnalysis.Operations.IDeclarationPatternOperation operation) Line 280	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.DeclarationPatternOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 7121	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement<Microsoft.CodeAnalysis.IOperation>(Microsoft.CodeAnalysis.IOperation operation) Line 25	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitArray<Microsoft.CodeAnalysis.IOperation>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations) Line 19	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.DefaultVisit(Microsoft.CodeAnalysis.IOperation operation) Line 47	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationVisitor.VisitPropertySubpattern(Microsoft.CodeAnalysis.Operations.IPropertySubpatternOperation operation) Line 8390	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.BasePropertySubpatternOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 7885	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement<Microsoft.CodeAnalysis.IOperation>(Microsoft.CodeAnalysis.IOperation operation) Line 25	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitArray<Microsoft.CodeAnalysis.IOperation>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations) Line 19	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.DefaultVisit(Microsoft.CodeAnalysis.IOperation operation) Line 47	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationVisitor.VisitRecursivePattern(Microsoft.CodeAnalysis.Operations.IRecursivePatternOperation operation) Line 8386	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.BaseRecursivePatternOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 7662	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement<Microsoft.CodeAnalysis.IOperation>(Microsoft.CodeAnalysis.IOperation operation) Line 25	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitArray<Microsoft.CodeAnalysis.IOperation>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations) Line 19	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.DefaultVisit(Microsoft.CodeAnalysis.IOperation operation) Line 47	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationVisitor.VisitIsPattern(Microsoft.CodeAnalysis.Operations.IIsPatternOperation operation) Line 8347	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.BaseIsPatternOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 5880	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement<Microsoft.CodeAnalysis.IOperation>(Microsoft.CodeAnalysis.IOperation operation) Line 25	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitArray<Microsoft.CodeAnalysis.IOperation>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations) Line 19	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.DefaultVisit(Microsoft.CodeAnalysis.IOperation operation) Line 47	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationVisitor.VisitUnaryOperator(Microsoft.CodeAnalysis.Operations.IUnaryOperation operation) Line 8312	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.BaseUnaryOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 4423	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement<Microsoft.CodeAnalysis.IOperation>(Microsoft.CodeAnalysis.IOperation operation) Line 25	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitArray<Microsoft.CodeAnalysis.IOperation>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations) Line 19	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.DefaultVisit(Microsoft.CodeAnalysis.IOperation operation) Line 47	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationVisitor.VisitConditional(Microsoft.CodeAnalysis.Operations.IConditionalOperation operation) Line 8314	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.BaseConditionalOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 4552	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitOperationArrayElement<Microsoft.CodeAnalysis.IOperation>(Microsoft.CodeAnalysis.IOperation operation) Line 25	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.VisitArray<Microsoft.CodeAnalysis.IOperation>(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations) Line 19	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.DefaultVisit(Microsoft.CodeAnalysis.IOperation operation) Line 47	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationVisitor.VisitBlock(Microsoft.CodeAnalysis.Operations.IBlockOperation operation) Line 8283	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.BaseBlockOperation.Accept(Microsoft.CodeAnalysis.Operations.OperationVisitor visitor) Line 2815	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Operations.OperationWalker.Visit(Microsoft.CodeAnalysis.IOperation operation) Line 36	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.Walker.Visit(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations, Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.AnalysisData analysisData, System.Threading.CancellationToken cancellationToken) Line 64	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.Walker.AnalyzeOperationsAndUpdateData(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.IOperation> operations, Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.AnalysisData analysisData, System.Threading.CancellationToken cancellationToken) Line 39	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageAnalysis.Run(Microsoft.CodeAnalysis.IOperation rootOperation, Microsoft.CodeAnalysis.ISymbol owningSymbol, System.Threading.CancellationToken cancellationToken) Line 35	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues.AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.AnalyzeUnusedValueAssignments(Microsoft.CodeAnalysis.Diagnostics.OperationBlockAnalysisContext context, bool isComputingUnusedParams, Microsoft.CodeAnalysis.PooledObjects.PooledHashSet<Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis.SymbolUsageResult> symbolUsageResultsBuilder, out bool hasBlockWithAllUsedSymbolWrites, out bool hasOperationNoneDescendant) Line 468	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues.AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.AnalyzeOperationBlockEnd(Microsoft.CodeAnalysis.Diagnostics.OperationBlockAnalysisContext context) Line 433	C#
```

### PR DESCRIPTION
**Pull request by mavasani**
…ons.GetValueUsageInfo

Fix #40499

### CODE REVIEW

File: src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs [hunk 1]
```diff
@@ -91,6 +91,14 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
                         //
                         return ValueUsageInfo.Write;
 
+                    case IPropertySubpatternOperation _:
+                        // A declaration pattern within a property sub-pattern is a
+                        // write for the declared local.
+                        // For example, 'x' is defined and assigned the value from 'obj.Property' below:
+                        //      if (obj is { Property : int x })
+                        //
+                        return ValueUsageInfo.Write;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: great comment.  was confused at first, but now i get it.   

for completeness, is `Property` recorded as a read somewhere?

File: src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs [hunk 2]
```diff
@@ -7415,6 +7415,60 @@ string M(object obj)
...
+        public async Task LocalUsedWithPropertySubPattern()
+        {
+            await TestDiagnosticMissingAsync(
+@"class C
+{
+    public object P { get; }
+    void M()
+    {
+        C [|c|] = new C();
+        var x = c is { P : int i };
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: can we have a tet with `var x` as well?

File: src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs (hunk 1 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: Yes, the IOperation tree has an `IPropertyReferenceOperation` for the reference to Property on left of the colon. This switch statement defaults it to a read operation. Doing a FAR on definition of `P` below does show this as a read reference:

```csharp
class C
{
    public object P { get; }
    void M(C c)
    {
        var x = c is { P: int i };
    }
}
```

image

File: src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs (hunk 2 above)
Comment (mavasani) [reply src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: Great suggestion! This case wasn't handled, so added a test and the fix for it.

**Comment by mavasani** [comment]
@dotnet/roslyn-ide for reviews - trivial fix for an assert.

**Comment by mavasani** [comment]
Thanks @sharwell @dpoeschl

**Pull request merged**